mayastor/src/bdev/dev/nvmx/controller.rs

```diff
@@ -236,155 +242,224 @@ impl<'a> NvmeController<'a> {
                     source: Errno::EBUSY,
                 });
             }
-            _ => {}
         }
 
+        info!(
+            "{} initiating controller reset, failover = {}",
+            self.name, failover
+        );
+
         if failover {
             warn!(
                 "{} failover is not supported for controller reset",
                 self.name
             );
         }
 
-        let reset_ctx = RESET_CTX_POOL
-            .get()
-            .unwrap()
-            .get(ResetCtx {
-                name: self.name.clone(),
-                cb,
-                cb_arg,
-                spdk_handle: self.ctrlr_as_ptr(),
-            })
-            .ok_or(CoreError::ResetDispatch {
-                source: Errno::ENOMEM,
-            })?;
+        let io_device = Arc::clone(&self.inner.as_ref().unwrap().io_device);
+        let reset_ctx = ResetCtx {
+            name: self.name.clone(),
+            cb,
+            cb_arg,
+            spdk_handle: self.ctrlr_as_ptr(),
+            io_device,
+        };
 
         // Mark controller as being under reset and schedule asynchronous reset.
-        self.set_state(NvmeControllerState::Resetting);
+        self.reset_active = true;
 
-        unsafe {
-            spdk_for_each_channel(
-                self.id as *mut c_void,
-                Some(NvmeController::reset_destroy_channels),
-                reset_ctx as *mut c_void,
-                Some(NvmeController::reset_destroy_channels_done),
-            );
-        }
+        // Iterate over all I/O channels and rrest/econfigure them one by one.
+        self.inner.as_ref().unwrap().io_device.traverse_io_channels(
+            NvmeController::_reset_destroy_channels,
+            NvmeController::_reset_destroy_channels_done,
+            reset_ctx,
+        );
         Ok(())
     }
 
-    fn complete_reset(reset_ctx: &ResetCtx, status: i32) {
-        // Set controller state to Running and invoke completion callback.
-        let c = NVME_CONTROLLERS
-            .lookup_by_name(&reset_ctx.name)
-            .expect("Controller was removed while reset is in progress");
-        let mut controller = c.lock().expect("lock poisoned");
+    fn _shutdown_channels(
```

Comment:
Is there a significance to these `_`'d functions?

Reply:
Nope.
I just wanted to denote callbacks. I will remove underscores.

Reply:
Ahhh! I think idiomatically `cb` is usually used for callbacks, so maybe `fn_name_cb`? https://stackoverflow.com/a/41081702